libraries/chain/smt_evaluator.cpp

```diff
@@ -161,32 +168,91 @@ void smt_setup_evaluator::do_apply( const smt_setup_operation& o )
 
+   if ( o.steem_units_min > 0 )
+   {
+      auto possible_hard_cap = util::smt::ico::steem_units_hard_cap( _db, o.symbol );
+
+      FC_ASSERT( possible_hard_cap.valid(),
+         "An SMT with a Steem Units Minimum of ${s} cannot succeed without an ICO tier.", ("s", o.steem_units_min) );
+
+      share_type hard_cap = *possible_hard_cap;
+
+      FC_ASSERT( o.steem_units_min <= hard_cap,
+         "The Steem Units Minimum must be less than the hard cap. Steem Units Minimum: ${s}, Hard Cap: ${c}",
+         ("s", o.steem_units_min)("c", hard_cap) );
+   }
+
    _db.modify( _token, [&]( smt_token_object& token )
    {
       token.phase = smt_phase::setup_completed;
       token.max_supply = o.max_supply;
    } );
 
-   const auto& token_ico = _db.create< smt_ico_object >( [&] ( smt_ico_object& token_ico_obj )
+   _db.create< smt_ico_object >( [&] ( smt_ico_object& token_ico_obj )
    {
       token_ico_obj.symbol = _token.liquid_symbol;
       token_ico_obj.contribution_begin_time = o.contribution_begin_time;
       token_ico_obj.contribution_end_time = o.contribution_end_time;
       token_ico_obj.launch_time = o.launch_time;
-      token_ico_obj.steem_units_soft_cap = o.steem_units_soft_cap;
-      token_ico_obj.steem_units_hard_cap = o.steem_units_hard_cap;
       token_ico_obj.steem_units_min = o.steem_units_min;
+      token_ico_obj.min_unit_ratio = o.min_unit_ratio;
+      token_ico_obj.max_unit_ratio = o.max_unit_ratio;
    } );
 
-   smt_setup_evaluator_visitor visitor( token_ico, _db );
-   o.initial_generation_policy.visit( visitor );
-
    smt_ico_launch_action ico_launch_action;
    ico_launch_action.control_account = _token.control_account;
    ico_launch_action.symbol = _token.liquid_symbol;
    _db.push_required_action( ico_launch_action, o.contribution_begin_time );
 }
 
+void smt_setup_ico_tier_evaluator::do_apply( const smt_setup_ico_tier_operation& o )
+{
+   FC_ASSERT( _db.has_hardfork( STEEM_SMT_HARDFORK ), "SMT functionality not enabled until hardfork ${hf}", ("hf", STEEM_SMT_HARDFORK) );
+
+   const smt_token_object& token = common_pre_setup_evaluation( _db, o.symbol, o.control_account );
+
+   smt_capped_generation_policy generation_policy;
+   smt_generation_policy_visitor< smt_capped_generation_policy > visitor( generation_policy );
+   o.generation_policy.visit( visitor );
+
+   if ( o.remove )
+   {
+      auto key = boost::make_tuple( token.liquid_symbol, o.steem_units_cap );
+      const auto* ito = _db.find< smt_ico_tier_object, by_symbol_steem_units_cap >( key );
+
+      FC_ASSERT( ito != nullptr,
+         "The specified ICO tier does not exist. Symbol: ${s}, Steem Units Cap: ${c}",
+         ("s", token.liquid_symbol)("c", o.steem_units_cap)
+      );
+
+      FC_ASSERT(
+         ito->symbol == o.symbol &&
+         ito->steem_units_cap == o.steem_units_cap &&
+         ito->capped_generation_policy.generation_unit.steem_unit == generation_policy.generation_unit.steem_unit &&
```

Comment:
Do we need to check the generation policy, or is a match on the symbol and cap sufficient enough? The reason to do so would be reduced impact on block size for removal of a generation policy.

Reply:
Symbol and cap is sufficient, but I thought it would be more consistent with `smt_token_emissions`, which require an exact match for removal.